From 14de22022d3fde05704d53e990f9d9798cd006c7 Mon Sep 17 00:00:00 2001
From: Zhengchao Shao <shaozhengchao@huawei.com>
Date: Thu, 30 Jun 2022 15:49:54 +0800
Subject: [PATCH 2670/4585] [Huawei] crypto: hisilicon/sec - use spin_lock_bh
 instead of spin_lock

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187109

--------------------------------

Commit cf0875b7d924 use spin_lock in sec2 driver, it will cause
soft lockup. Use spin_lock_bh instead of spin_lock.

watchdog: BUG: soft lockup - CPU#30 stuck for 23s!
[kworker/u262:0:953]
Workqueue: 0000:76:00.0 qm_work_process [hisi_qm]
pc : native_queued_spin_lock_slowpath+0x14c/0x324
lr : sec_alloc_req_id+0xd8/0xf4 [hisi_sec2]
sp : ffff8000100f39a0
x29: ffff8000100f39a0 x28: ffff42d657606800
x27: ffff42d657606800 x26: 0000000000000008
x25: 00000000000005b8 x24: 00000000000005c0
x23: ffff42f5838a4780 x22: ffffb4ff94899c90
x21: ffff42d657606b00 x20: ffff42f5a54a4078
x19: ffff42f5a5492050 x18: 0000000000000000
x17: 0000000000000000 x16: ffffb4ff93472480
x15: 0000000000000004 x14: 342d552461874163
x13: e54a9a9205ae275d x12: 1248650e689d596b
x11: 0000000000000000 x10: 00000000000005c0
x9 : ffffb4ff6e82a288 x8 : 000000000000ffff
x7 : 0000000000000010 x6 : 0000000000000010
x5 : 0000000000000000 x4 : 0000000000012050
x3 : ffff42f5a54a4078 x2 : ffffb4ff9301ff90
x1 : 0000000000000000 x0 : 0000000000000101
Call trace:
native_queued_spin_lock_slowpath+0x14c/0x324
sec_request_init+0x7c/0x14c [hisi_sec2]
sec_process+0x28/0x1ac [hisi_sec2]
sec_aead_decrypt+0x64/0xa0 [hisi_sec2]
crypto_aead_decrypt+0x3c/0x60
crypto_rfc4106_decrypt+0x34/0x40
crypto_aead_decrypt+0x3c/0x60
seqiv_aead_decrypt+0x90/0xb0
crypto_aead_decrypt+0x3c/0x60
esp_input+0x1d8/0x320 [esp4]
xfrm_input+0x9b8/0xb00
xfrm4_rcv+0x3c/0x50
xfrm4_esp_rcv+0x48/0x94
ip_protocol_deliver_rcu+0xf4/0x200
ip_local_deliver_finish+0x58/0x70
ip_local_deliver+0x68/0x120
ip_rcv_finish+0x98/0xc0
ip_rcv+0x58/0xe4
__netif_receive_skb_one_core+0x60/0x90
__netif_receive_skb+0x20/0x70
process_backlog+0xa8/0x1d4
napi_poll+0xcc/0x264
net_rx_action+0xd4/0x21c
__do_softirq+0x130/0x358
irq_exit+0x11c/0x13c
__handle_domain_irq+0x88/0xf0
gic_handle_irq+0x78/0x2c0
el1_irq+0xb8/0x140
__radix_tree_lookup+0x7c/0xbc
idr_remove+0x1c/0x2c
sec_free_req_id+0x80/0xe4 [hisi_sec2]
sec_aead_callback+0x6c/0x150 [hisi_sec2]
sec_req_cb+0x108/0x1f0 [hisi_sec2]
qm_poll_qp+0x100/0x190 [hisi_qm]
qm_work_process+0x88/0x110 [hisi_qm]
process_one_work+0x1d0/0x490
worker_thread+0x158/0x3d0
kthread+0x108/0x13c
ret_from_fork+0x10/0x18

Fixes: 2f32b51b609f ("crypto: hisilicon/sec - don't sleep when in softirq")
Signed-off-by: Zhengchao Shao <shaozhengchao@huawei.com>
---
 drivers/crypto/hisilicon/sec2/sec_crypto.c | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

diff --git a/drivers/crypto/hisilicon/sec2/sec_crypto.c b/drivers/crypto/hisilicon/sec2/sec_crypto.c
index bf58656405e9..b2228a935dcd 100644
--- a/drivers/crypto/hisilicon/sec2/sec_crypto.c
+++ b/drivers/crypto/hisilicon/sec2/sec_crypto.c
@@ -127,11 +127,11 @@ static int sec_alloc_req_id(struct sec_req *req, struct sec_qp_ctx *qp_ctx)
 {
 	int req_id;
 
-	spin_lock(&qp_ctx->req_lock);
+	spin_lock_bh(&qp_ctx->req_lock);
 
 	req_id = idr_alloc_cyclic(&qp_ctx->req_idr, NULL,
 				  0, QM_Q_DEPTH, GFP_ATOMIC);
-	spin_unlock(&qp_ctx->req_lock);
+	spin_unlock_bh(&qp_ctx->req_lock);
 	if (unlikely(req_id < 0)) {
 		dev_err(req->ctx->dev, "alloc req id fail!\n");
 		return req_id;
@@ -156,9 +156,9 @@ static void sec_free_req_id(struct sec_req *req)
 	qp_ctx->req_list[req_id] = NULL;
 	req->qp_ctx = NULL;
 
-	spin_lock(&qp_ctx->req_lock);
+	spin_lock_bh(&qp_ctx->req_lock);
 	idr_remove(&qp_ctx->req_idr, req_id);
-	spin_unlock(&qp_ctx->req_lock);
+	spin_unlock_bh(&qp_ctx->req_lock);
 }
 
 static u8 pre_parse_finished_bd(struct bd_status *status, void *resp)
@@ -273,7 +273,7 @@ static int sec_bd_send(struct sec_ctx *ctx, struct sec_req *req)
 	    !(req->flag & CRYPTO_TFM_REQ_MAY_BACKLOG))
 		return -EBUSY;
 
-	spin_lock(&qp_ctx->req_lock);
+	spin_lock_bh(&qp_ctx->req_lock);
 	ret = hisi_qp_send(qp_ctx->qp, &req->sec_sqe);
 
 	if (ctx->fake_req_limit <=
@@ -281,10 +281,10 @@ static int sec_bd_send(struct sec_ctx *ctx, struct sec_req *req)
 		list_add_tail(&req->backlog_head, &qp_ctx->backlog);
 		atomic64_inc(&ctx->sec->debug.dfx.send_cnt);
 		atomic64_inc(&ctx->sec->debug.dfx.send_busy_cnt);
-		spin_unlock(&qp_ctx->req_lock);
+		spin_unlock_bh(&qp_ctx->req_lock);
 		return -EBUSY;
 	}
-	spin_unlock(&qp_ctx->req_lock);
+	spin_unlock_bh(&qp_ctx->req_lock);
 
 	if (unlikely(ret == -EBUSY))
 		return -ENOBUFS;
@@ -1382,7 +1382,7 @@ static struct sec_req *sec_back_req_clear(struct sec_ctx *ctx,
 {
 	struct sec_req *backlog_req = NULL;
 
-	spin_lock(&qp_ctx->req_lock);
+	spin_lock_bh(&qp_ctx->req_lock);
 	if (ctx->fake_req_limit >=
 	    atomic_read(&qp_ctx->qp->qp_status.used) &&
 	    !list_empty(&qp_ctx->backlog)) {
@@ -1390,7 +1390,7 @@ static struct sec_req *sec_back_req_clear(struct sec_ctx *ctx,
 				typeof(*backlog_req), backlog_head);
 		list_del(&backlog_req->backlog_head);
 	}
-	spin_unlock(&qp_ctx->req_lock);
+	spin_unlock_bh(&qp_ctx->req_lock);
 
 	return backlog_req;
 }
-- 
2.17.1

